File: python/src/contriwork_market_data/_adapter.py
"""Adapter protocol — INTERNAL. Adapter authors implement this.

CONTRACT.md §4. Every adapter (CoinGecko, Binance, …) must implement these
four operations plus ``capability`` and ``adapter_id``. The orchestrator
(:class:`~contriwork_market_data.client.MarketDataClient`) wraps adapter
calls with cache, rate limiting, and chain fallback so adapters can stay
narrow and provider-focused.

Adapter conformance is structural (``Protocol``), so test fakes and stub
adapters do not need to subclass anything — they just need matching
attributes and method signatures.
"""

from __future__ import annotations

from collections.abc import AsyncIterator
from datetime import datetime
from typing import Protocol, runtime_checkable

from .types import Candle, Capability, Interval, OrderBook, SpotPrice, Ticker

__all__ = ["Adapter"]


@runtime_checkable
class Adapter(Protocol):
    adapter_id: str
    capability: Capability

    async def get_spot(self, symbol: str, quote_currency: str) -> SpotPrice: ...

    async def get_ohlcv(
        self,
        symbol: str,
        interval: Interval,
        since: datetime | None,
        limit: int,
    ) -> list[Candle]: ...

    async def get_order_book(self, symbol: str, depth: int) -> OrderBook: ...

    def subscribe_ticker(self, symbol: str) -> AsyncIterator[Ticker]: ...

File: typescript/src/adapter.ts
/**
 * Adapter contract — CONTRACT.md §4. Each provider adapter implements this;
 * the orchestrator (`MarketDataClient`) layers cache, rate limiting, and
 * chain fallback on top.
 */
import type {
  Candle,
  Capability,
  Interval,
  OrderBook,
  SpotPrice,
  Ticker,
} from "./types.js";

/** A provider adapter. */
export interface MarketDataAdapter {
  /** Stable kebab-case adapter id (e.g. `"coingecko"`). */
  readonly adapterId: string;

  /** Static description of what the adapter supports. */
  readonly capability: Capability;

  /** Fetch the latest spot price. */
  getSpot(
    symbol: string,
    quoteCurrency: string,
    signal?: AbortSignal,
  ): Promise<SpotPrice>;

  /** Fetch historical OHLCV candles. */
  getOhlcv(
    symbol: string,
    interval: Interval,
    since: Date | undefined,
    limit: number,
    signal?: AbortSignal,
  ): Promise<readonly Candle[]>;

  /** Fetch the top-N order book. */
  getOrderBook(
    symbol: string,
    depth: number,
    signal?: AbortSignal,
  ): Promise<OrderBook>;

  /** Open a native ticker stream for the symbol. */
  subscribeTicker(symbol: string, signal?: AbortSignal): AsyncIterable<Ticker>;
}

/** A lazy credential provider — resolves a key on demand. */
export type ApiKeyProvider = (
  signal?: AbortSignal,
) => Promise<string | undefined>;
